### PR TITLE
[Editor] Add some telemetry to know how often the editing features are used (bug 1782254)

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

**Review by Snuffleupagus — COMMENTED** [review]
I can't help but wonder if we could perhaps reduce the amount of code needed, primarily in the `src/display/editor/`-folder, in order pass the necessary editor-type onto the viewer?

More specifically, in *addition* to the code that you're adding to handle saving/printing, we might be able to use the `AnnotationStorage` along these lines:
```diff
diff --git a/src/display/annotation_storage.js b/src/display/annotation_storage.js
index 6b558f8f0..eb52a4823 100644
--- a/src/display/annotation_storage.js
+++ b/src/display/annotation_storage.js
@@ -31,6 +31,8 @@ class AnnotationStorage {
     // can have undesirable effects.
     this.onSetModified = null;
     this.onResetModified = null;
+
+    this.onAnnotationEditor = null;
   }
 
   /**
@@ -100,6 +102,13 @@ class AnnotationStorage {
     if (modified) {
       this.#setModified();
     }
+
+    if (
+      value instanceof AnnotationEditor &&
+      typeof this.onAnnotationEditor === "function"
+    ) {
+      this.onAnnotationEditor(value.constructor._type);
+    }
   }
 
   /**
diff --git a/web/app.js b/web/app.js
index 8b4e6265e..fd05a99aa 100644
--- a/web/app.js
+++ b/web/app.js
@@ -1734,6 +1734,12 @@ const PDFViewerApplication = {
         delete this._annotationStorageModified;
       }
     };
+    annotationStorage.onAnnotationEditor = typeStr => {
+      this.externalServices.reportTelemetry({
+        type: "editing",
+        data: { type: typeStr },
+      })
+    };
   },
 
   setInitialView(
```

File: src/display/editor/freetext.js [hunk 1]
```diff
@@ -63,6 +63,8 @@ class FreeTextEditor extends AnnotationEditor {
     ],
   ]);
 
+  static _type = "freetext";
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/display/editor/freetext.js:66]: I thought about using `AnnotationEditorType`, but if for any reason we change something in those numbers we'll have to think about changing them in the jsm in m-c in order to get the correct bucket, hence I thought that having this name should be more stable over the time.

[the remark this answers is not in the document]
Comment (calixteman) [reply src/display/editor/freetext.js:66]: That said, I've no plan to change those numbers and the telemetry should be there for only few months...

[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/display/editor/freetext.js:66]: > I thought about using `AnnotationEditorType`, but if for any reason we change something in those numbers we'll have to think about changing them in the jsm in m-c in order to get the correct bucket, hence I thought that having this name should be more stable over the time.
>
> That said, I've no plan to change those numbers and the telemetry should be there for only few months...

In that case, using strings seem fine to me. (Even more so, if this code is only temporary anyway.)

**Comment by calixteman** [comment]
> I can't help but wonder if we could perhaps reduce the amount of code needed, primarily in the `src/display/editor/`-folder, in order pass the necessary editor-type onto the viewer?
> 
> More specifically, in _addition_ to the code that you're adding to handle saving/printing, we might be able to use the `AnnotationStorage` along these lines:
> 
> ```diff
> diff --git a/src/display/annotation_storage.js b/src/display/annotation_storage.js
> index 6b558f8f0..eb52a4823 100644
> --- a/src/display/annotation_storage.js
> +++ b/src/display/annotation_storage.js
> @@ -31,6 +31,8 @@ class AnnotationStorage {
>      // can have undesirable effects.
>      this.onSetModified = null;
>      this.onResetModified = null;
> +
> +    this.onAnnotationEditor = null;
>    }
>  
>    /**
> @@ -100,6 +102,13 @@ class AnnotationStorage {
>      if (modified) {
>        this.#setModified();
>      }
> +
> +    if (
> +      value instanceof AnnotationEditor &&
> +      typeof this.onAnnotationEditor === "function"
> +    ) {
> +      this.onAnnotationEditor(value.constructor._type);
> +    }
>    }
>  
>    /**
> diff --git a/web/app.js b/web/app.js
> index 8b4e6265e..fd05a99aa 100644
> --- a/web/app.js
> +++ b/web/app.js
> @@ -1734,6 +1734,12 @@ const PDFViewerApplication = {
>          delete this._annotationStorageModified;
>        }
>      };
> +    annotationStorage.onAnnotationEditor = typeStr => {
> +      this.externalServices.reportTelemetry({
> +        type: "editing",
> +        data: { type: typeStr },
> +      })
> +    };
>    },
>  
>    setInitialView(
> ```

It's definitely more clever than my solution but it adds a kind of invisible dependency between `annotationStorage` and `externalServices`, is it really something we want ?

**Comment by Snuffleupagus** [comment]
> but it adds a kind of invisible dependency between `annotationStorage` and `externalServices`,

To me it's not that much different from the pre-existing callback-functions available in the `AnnotationStorage` class, and in a way the new `hasAnnotationEditors` method (which we still need of course) also introduces a kind-of dependency here.
The advantage of the suggested approach, at least in my opinion, is that it's smaller and easier to overview which hopefully means that it's less likely to break with any future changes to the Editing-functionality.

**Pull request merged**